[LLDB] Update DIL handling of array subscripting.

cmtice opened the pull request:
This updates the DIL code for handling array subscripting to more closely match and handle all the cases from the original 'frame var' implementation. Also updates the DIL array subscripting test. This particularly fixes some issues with handling synthetic children, objc pointers, and accessing specific bits within scalar data types.

cmtice committed:
[LLDB] Update DIL handling of array subscripting.

This updates the DIL code for handling array subscripting to more
closely match and handle all the cases from the original
'frame var' implementation. Also updates the DIL array subscripting
test. This particularly fixes some issues with handling
synthetic children, objc pointers, and accessing specific bits
within scalar data types.

labath reviewed (commented):
I see the bitfield handling is reflected in the test case, but I don't see anything about handling of synthetic children or objc pointers. For the objc case, I'm very tempted to *not* copy that bit of code over, since it: a) is untested (or you wouldn't have been able to start using the new implementation); b) doesn't fit into our concept of being language-agnostic.

It looks like that's mainly error handling code guarding a call to GetSyntheticArrayMember. That function eventually calls `GetCompilerType().GetChildCompilerTypeAtIndex`, so I think that special handling of ObjC should happen there.

That would also reduce the maze of the if statements, which is making the function hard to review.

cmtice committed:
Address reviewer comments:
- Remove code testing for objc pointers
- No else-after-return
- Rename errMsg err_msg
- Add child_idx+1 argument to calls to GetNumChildrenIgnoringErrors
- Fix typo.

cmtice committed:
Fix clang-format issues.

Michael137 commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -330,40 +330,85 @@ Interpreter::Visit(const ArraySubscriptNode *node) {
+          "array index {0} is not valid for \"({1}) {2}\"", child_idx,
+          base->GetTypeName().AsCString("<invalid type>"),
+          var_expr_path_strm.GetData());
+      return llvm::make_error<DILDiagnosticError>(m_expr, err_msg,
```
You can `std::move` these `err_msg`s

cmtice replied:
Done

cmtice committed:
Add std::move for err_msg strings and dynamic value.
Add test case for GetSyntheticArrayMember.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -330,40 +330,85 @@ Interpreter::Visit(const ArraySubscriptNode *node) {
+      lldb::ValueObjectSP dynamic_value_sp(
+          child_valobj_sp->GetDynamicValue(m_use_dynamic));
+      if (dynamic_value_sp)
+        child_valobj_sp = dynamic_value_sp;
```
Done.

cmtice committed:
Fix clang format issues.

cmtice commented:
I've addressed all the review comments and added the test case. Please review again when you have a few minutes. Thanks!

rastogishubham commented:
Hi, it seems like this change broke greendragon

https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/31450/

It was reverted with cd0bf2735bcd1e9a21dd10169782060a3702c447
